### PR TITLE
A0-3391: Safer own block broadcast

### PR DESCRIPTION
**Pull request by lesniak43**
# Description

We broadcast own blocks only after they are imported.

## Type of change

- Bug fix (non-breaking change which fixes an issue)

# Checklist:

- I have added tests
- I have created new documentation

### CODE REVIEW

File: finality-aleph/src/block/substrate/verification/cache.rs [hunk 1]
```diff
@@ -403,6 +403,16 @@ where
             maybe_equivocation_proof,
         })
     }
+
+    fn own_block(&self, header: &Header) -> bool {
+        // on average: ~2000 elements in the cache, one check per second
+        for (cached_header, certainly_own) in self.equivocation_cache.values() {
```
Comment (timorleph) [inline finality-aleph/src/block/substrate/verification/cache.rs:409]: Come on, don't loop, add some indexing which makes this check O(1).

**Review by timorleph — APPROVED** [review]
Looks ok, just one comment.

File: finality-aleph/src/block/substrate/verification/cache.rs (hunk 1 above)
Comment (woocash2) [reply finality-aleph/src/block/substrate/verification/cache.rs:409]: Agree, I guess one way to do that is to convert `equivocation_cache` into `HashMap<Slot, H>` and also keep `certainly_own: HashSet<_>` where you insert `header.hash()` of each certainly own `header`.

Comment (lesniak43) [reply finality-aleph/src/block/substrate/verification/cache.rs]: fixed

File: finality-aleph/src/sync/handler/mod.rs [hunk 2]
```diff
@@ -505,7 +505,7 @@ where
     pub fn block_imported(
         &mut self,
         header: J::Header,
-    ) -> Result<(), <Self as HandlerTypes>::Error> {
+    ) -> Result<Option<Vec<ResponseItem<B, J>>>, <Self as HandlerTypes>::Error> {
```
[the remark this answers is not in the document]
Comment (lesniak43) [reply finality-aleph/src/sync/handler/mod.rs]: fixed

File: finality-aleph/src/sync/handler/mod.rs [hunk 3]
```diff
@@ -861,9 +868,11 @@ where
     pub fn handle_own_block(
```
[the remark this answers is not in the document]
Comment (lesniak43) [reply finality-aleph/src/sync/handler/mod.rs:869]: fixed

**Pull request merged**